src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -178,8 +182,17 @@ public SCMFileSystem build(@NonNull SCMSource source, @NonNull SCMHead head, @Ch
                 } else if (pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.HEAD) {
                     ref = "pull-requests/" + pr.getId() + "/from";
                 } else if (pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) {
-                    // No longer supported since bitbucket server v7, falls back to heavycheckout
-                    ref = "pull-requests/" + pr.getId() + "/merge";
+                    // Bitbucket server v7 doesn't support `merge` endpoint
+                    // We don't return `ref` when working with v7
+                    // so that pipeline steps retrieving single files fall back to heavyweight checkout properly
+                    AbstractBitbucketEndpoint endpointConfig = BitbucketEndpointConfiguration.get().findEndpoint(src.getServerUrl());
+                    final BitbucketServerEndpoint endpoint = endpointConfig instanceof BitbucketServerEndpoint ?
+                            (BitbucketServerEndpoint) endpointConfig : null;
+                    if (endpoint != null) {
+                        if(endpoint.getServerVersion() != BitbucketServerVersion.VERSION_7) {
+                            ref = "pull-requests/" + pr.getId() + "/merge";
+                        }
+                    }
```

Comment:
Looking at line 177 above it seems we should return `null` rather than a `ref = null`. 

This is where the fallback occurs, I believe. 
https://github.com/jenkinsci/scm-api-plugin/blob/767f59ed48d68e73d7d3a34a0ca85e96df874b30/src/main/java/jenkins/scm/api/SCMFileSystem.java#L197-L205 



```suggestion
                    // Bitbucket server v7 doesn't support `merge` endpoint
                    // We don't return `ref` when working with v7
                    // so that pipeline steps retrieving single files fall back to heavyweight checkout properly
                    AbstractBitbucketEndpoint endpointConfig = BitbucketEndpointConfiguration.get().findEndpoint(src.getServerUrl());
                    final BitbucketServerEndpoint endpoint = endpointConfig instanceof BitbucketServerEndpoint ?
                            (BitbucketServerEndpoint) endpointConfig : null;
                    if (endpoint != null && endpoint.getServerVersion() != BitbucketServerVersion.VERSION_7) {
                        ref = "pull-requests/" + pr.getId() + "/merge";
                    } else {
                        // returning null 
                        return null;
                    }
```